File: csharp/817.linked-list-components.cs
/*
 * @lc app=leetcode id=817 lang=csharp
 *
 * [817] Linked List Components
 */

using System.Collections.Generic;

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int NumComponents(ListNode head, int[] G) {
        var set = new HashSet<int>(G);
        int ret = 0;
        bool p = false;
        while (head != null) {
            if (set.Contains(head.val)) {
                if (!p) {
                    ++ret;
                }
                p = true;
            } else {
                p = false;
            }
            head = head.next;
        }
        return ret;
    }
}
// @lc code=end

File: java/817.linked-list-components.java
/*
 * @lc app=leetcode id=817 lang=java
 *
 * [817] Linked List Components
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] G) {
        Set<Integer> set = Arrays.stream(G).boxed().collect(Collectors.toSet());
        int ret = 0;
        while (head != null) {
            if (set.contains(head.val) && (head.next == null || !set.contains(head.next.val))) {
                ++ret;
            }
            head = head.next;
        }
        return ret;
    }
}
// @lc code=end

